### DISCUSSION TITLE
Using a dependency only when needed instead of everytime

### DISCUSSION THREAD

**Discussion by InterStella0** [fastapi/fastapi#11206] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from sqlmodel.ext.asyncio.session import AsyncSession
from sqlalchemy.ext.asyncio import create_async_engine
from fastapi import FastAPI, Depends

app = FastAPI()
app.state.models_cached = {}

def create_model_from_store(data):
   return {'id': 'some_id'}
def engine():
   return create_async_engine('url')

async def get_session():
    async with AsyncSession(engine()) as session:
        yield session

async def get_model(*,
   model_id: str, 
   session: AsyncSession = Depends(get_session), 
   request: fastapi.Request
):
    models_cached = request.app.state.models_cached
    if model := models_cached.get(model_id):
        return model
    if not (data := await session.get(Model, model)):
        model = create_model_from_store(data)
        models_cached[model_id] = model
    return model

@app.get('/models/{model_id}/')
def get_some_model(model = Depends(get_model)):
    ...
```

## Context
So what i'm trying to do is to cache the Model after it has been requested once. It's to avoid the need to retrieve the data from the database multiple time from **multiple requests**. This is because this endpoint gets called multiple times a second. So creating a session for every request slows it down significantly. I'm aware that dependency gets cached for a singular request.

## The problem: 

Using Depends essentially requires it to call `get_session` every requests making the caching mechanism useless.

### Operating System

Windows

### FastAPI Version

0.109.0

### Pydantic Version

2.5.3

### Python Version

Python 3.10.1

### Additional Context

 I have a way to avoid this which is to not use Depends like so 
```py
async def get_model(*,
   model_id: str, 
   request: fastapi.Request
):
    models_cached = request.app.state.models_cached
    if model := models_cached.get(model_id):
        return model
    async with AsyncSession(engine()) as session:
        if not (data := await session.get(Model, model)):
            model = create_model_from_store(data)
            models_cached[model_id] = model
    return model
```

This works, but now I have 2 separate sources of creating sessions which does not play well with unittest/pytest. I basically have a separate database specifically for testing, so replacing `get_session` is easily done via  `dependency_overrides`.
```py
@pytest.fixture
def client_authenticated_memory(memory_engine):
    async def memory_session():
        async with AsyncSession(memory_engine) as session:
            yield session

    app.dependency_overrides[database.get_session] = memory_session
    return TestClient(app)
```
But now `get_model` also requires to be replaced. So I need a way to only override get_session without needing to override other dependencies.

I wish there was a mechanism where we can 'acquire' the dependency whenever we want instead of loading it at the top of the param. Something like this.
```py
async def get_model(*,
   model_id: str, 
   request: fastapi.Request,
   ac_session: Acquire[AsyncSession] = Depends(get_session, acquire=True)
   ):
   ... # caching algorithm
   session = await ac_session.acquire()
```

**Comment by chrisK824** [comment]
Have you tried a decorator on top ot the endpoint route? My guess is that you could try your cache check in there and return the result if so. Otherwise just proceed with the handler. Ofc, the dummy solution is to make the cache query in the dependency handler itself and if result is found then yield before creating a database session.

**Reply by InterStella0** [reply]
can you elaborate on this? I don't see how I can implement it with just a decorator since I don't know how fastapi would behave

**Comment by YuriiMotov** [comment, marked as answer]
A couple of ways to solve this:
1. Use `fastapi-cache2` or similar library to cache endpoint's result. It will cache results for each combination of request path and parameters.
2. Create `get_sessionmaker` dependency and use it as subdependency in `get_session`:
```
engine = create_async_engine('url')
async_session_maker = async_sessionmaker(engine, expire_on_commit=False)

def get_sessionmaker():
    return async_session_maker

async def get_session(sessionmaker: async_sessionmaker = Depends(get_sessionmaker)):
    async with sessionmaker() as session:
        yield session
```
In get_model you can use `get_sessionmaker` dependency and request session only if you need.
And in tests you can just override `get_sessionmaker` dependency.

But actually, as I know it's not so expensive task to get DB session in SQLAlcjemy, since it only does expensive work when you use this session (by calling commit, get, execute, scalar or scalars methods).

Also, keep in mind that if you run your app using guvicorn with multiple workers, every worker will have it's own memory.

**Reply by InterStella0** [reply]
seems like the second one is what I need since I like the idea, I don't really need another library just for caching. I've tried not to overengineer get_model by not using caching, but getting session every request was slow and significant enough that I needed them. The request goes several hundred per second, albeit not constantly.
